### ISSUE TITLE
Temporary tables for common, time-consuming queries

### ISSUE THREAD

**Issue by wjchulme** [opensafely/cohort-extractor#438]
This has been discussed on slack previously, mainly for ethnicity but it may be useful for other variables too.  

The general idea is to build a new ethnicity table every time new `S1` data is imported to the database. Extracting ethnicity is quite slow -- with an ethnicity table the time-consuming query on each patient's entire GP record only has to happen once per `S1` update, rather than every time a `generate_cohort` is run. Given ethnicity is in most studies it could save quite a bit of time.

The table would just contain the latest known ethnicity code, with [ID, ethnicity6, ethnicity16], and the function would be

```
ethnicity16 = patients.ethnicity(
    grouping = 16,
    return_expectations = ...
)
```

Since ethnicity (almost always) doesn't change, this approach would suit most cases. If you really want to know how ethnicity would have been calculated at some point in the past, then you can use the old way with `patients.with_these_clinical_events()`.

From @evansd:

> In terms of workflow it would probably be easiest for us to to this ourselves [not TPP] and write them to `OPENCoronaTempTables`

From @alexwalkercebm:
> For smoking, could we cache the output of the query every time a new date gets used? Then the query for each date should only need to be run once

**Comment by sebbacon** [comment]
Better, we'd introduce a caching mechanism so this happens transparently for the user.  This would tie into our idea of a "variables library". If you use a canned ethnicity query then we can cache that for you at build time.

**Comment by wjchulme** [comment]
Would this cache across different workspaces with different study populations?

**Comment by evansd** [comment]
> Would this cache across different workspaces with different study populations?

Yes, I imagine it would cache the results for the entire population and then each study would only use the results for its specific cohort.

> Would it mean that the first extract each week (per user? per canned query?) is slow?

Yes, the first time each query is run (not per user) would be slow. It would be possible to automatically run these queries at the beginning of the week ("cache warming" is the term) so the slowness doesn't effect any particular user.
